modules/ldap_query/lambda/lambda.py

```diff
@@ -0,0 +1,467 @@
+            Key=dest_object_name,
+            Body=object_data
+            )
+    except s3.exceptions.ClientError as e:
+        # AllAccessDisabled error == bucket not found
+        # NoSuchKey or InvalidRequest
+        # error == (dest bucket/obj == src bucket/obj)
+        log.error(e)
+        return False
+    finally:
+        if isinstance(src_data, str):
+            object_data.close()
+    return True
+
+
+def create_presigned_url(bucket_name, object_name, expiration=3600):
+    s3 = boto3.client('s3')
+    try:
+        response = s3.generate_presigned_url(
+            'get_object',
+            Params={
+                'Bucket': bucket_name,
+                'Key': object_name
+                },
+            ExpiresIn=expiration
+        )
+    except s3.exceptions.ClientError as e:
+        log.error(e)
+        return None
+    # The response contains the presigned URL
+    return response
+
+
+def upload_artifact(artifact):
+    """Uploads an artifact to s3 and generates a presigned url
+
+    Arguments:
+        artifact {dict} -- dictionary containing the artifact's
+        contents and file name
+
+    Returns:
+        dict -- dictionary containing the object name and presigned url
+    """
+    bucket_name = os.environ['ARTIFACTS_BUCKET']
+    log.debug(f'Uploading object: {artifact["file_name"]} to {bucket_name}')
+    if put_object(
+            bucket_name,
+            artifact["file_name"],
+            artifact["content"].encode("utf-8")):
+        presigned_url = create_presigned_url(
+            bucket_name, artifact["file_name"])
+    else:
+        log.error('Encountered error when uploading artifact')
+    is_raw_scan_result = False
+    if artifact.get("raw_scan_results"):
+        is_raw_scan_result = True
+    return {
+        "file_name": artifact["file_name"],
+        "url": presigned_url,
+        "raw_scan_results": is_raw_scan_result
+    }
+
+
+def upload_all_artifacts(content):
+    artifacts = generate_artifacts(content)
+    log.debug(f"generated artifacts: {artifacts}")
+    response = []
+    for artifact in artifacts:
+        # log.info(artifacts[artifact])
+        response.append(upload_artifact(artifacts[artifact]))
+    return response
+
+
+def get_user_counts(users):
+    response = {}
+    for key in users:
+        response[key] = len(users[key])
+    return response
+
+
+def retrieve_s3_object_contents(
+    s3_obj,
+    bucket=os.environ['ARTIFACTS_BUCKET']
+):
+    return json.loads(s3.get_object(
+        Bucket=bucket,
+        Key=s3_obj
+        )['Body'].read().decode('utf-8'))
+
+
+def handler(event, context):
+    """
+    expected event:
+    {
+        "action": query | disable
+    }
+    """
+    log.info(f'Received event: {event}')
+    if event.get("Payload"):
+        event = event['Payload']
+    elif event.get("Input"):
+        event = event["Input"]
+    # try:
+    if event.get("action"):
+        if event['action'] == "query":
```

Comment:
i'd recommend creating two more functions, `query_handler` and `disable_handler` and moving the logic there. e.g.

```
def query_handler():
            users = LdapMaintainer().get_stale_users()
            log.debug(f"Ldap query results: {users}")
            return {
                "query_results": {
                    "totals": get_user_counts(users)
                },
                "artifacts": upload_all_artifacts(users)
                }
```

Reply:
You can then setup a strategy pattern to handle the event:

```
strategy = {
    'query': query_handler
    'disable': disable_handler
}

# call the action handler, let it raise on invalid/missing action
strategy[event['action']]()
```

Reply:
I _think_ I reworked this. I first tired the strategy map but when I defined the functions as key values python tried to run each function. Is there a syntax or OOO thing I'm missing?